### PR TITLE
HAWQ-1226. HAWQ core dump due to enable ranger while RPS is down 

### PR DESCRIPTION
**Pull request by interma**
also include:
1. fix core dump while RPS not running
2. fix libcurl write_callback() problem
3. add libcurl timeout (hard-code 30s now)
4. fix some compile warnings
5. fix compile error: add json.h to checkinc.py

### CODE REVIEW

File: src/backend/libpq/be-secure.c [hunk 1]
```diff
@@ -329,6 +331,7 @@ report_commerror(const char *err_msg)
 
 	RESUME_INTERRUPTS();
 }
+*/
```
Comment (stanlyxiang) [inline src/backend/libpq/be-secure.c]: If it is unused now, could we remove it directly  ?

File: src/backend/libpq/rangerrest.c [hunk 2]
```diff
@@ -245,279 +257,326 @@ json_object *create_ranger_request_json_batch(List *args)
...
+	memcpy(curl->response.buffer + curl->response.size, contents, realsize);
+	curl->response.size += realsize;
+	curl->response.buffer[curl->response.size] = '\0';
+	elog(LOG, "read from Ranger Restful API: %s", curl->response.buffer);
+
+	return realsize;
 }
 
-void call_ranger_rest(CURL_HANDLE curl_handle, const char* request)
+
```
Comment (stanlyxiang) [inline src/backend/libpq/rangerrest.c]: too many blank line.

**Comment by interma** [comment]
@linwen @xunzhang @stanlyxiang @zhangh43 help to review, thanks.

1. fix core dump while RPS not running
rangerrest.c:435 

2. fix libcurl write_callback() problem
rangerrest.c:374

3. add libcurl timeout (hard-code 30s now)
rangerrest.c:451

4. fix some compile warnings
5. fix compile error: add json.h to checkinc.py

File: src/backend/libpq/rangerrest.c (hunk 2 above)
Comment (interma) [inline src/backend/libpq/rangerrest.c]: Tks!
I will fix it in next PR by me or others.

File: src/backend/libpq/be-secure.c (hunk 1 above)
Comment (interma) [inline src/backend/libpq/be-secure.c]: I am not sure, maybe someone added it for future...

File: src/test/regress/checkinc.py [hunk 3]
```diff
@@ -78,9 +78,10 @@
     'winsock.h':         [],
     'winsock2.h':        [],
     'ws2tcpip.h':        [],
-    'hdfs/hdfs.h': 		 [],
+    'hdfs/hdfs.h': 	 [],
     'quicklz1.h':	 [],
     'quicklz3.h':	 [],
+    'json-c/json.h':     [],
```
Comment (xunzhang) [inline src/test/regress/checkinc.py]: Does `checkinc.py` apply only when executing `make install`? I think this fix is necessary but I don't have a Mac environment to test it.

File: src/backend/libpq/rangerrest.c (hunk 2 above)
Comment (xunzhang) [inline src/backend/libpq/rangerrest.c]: You don't need to send another PR, you can append further commits in this PR.

File: src/backend/libpq/rangerrest.c [hunk 4]
```diff
@@ -24,60 +24,72 @@
...
-    }
-    return RANGERCHECK_OK;
+	Assert(buffer != NULL);
+	if (strlen(buffer) == 0)
+		return RANGERCHECK_UNKNOWN;
+
+	elog(LOG, "read from Ranger Restful API: %s", buffer);
+
+	struct json_object *response = json_tokener_parse(buffer);
+	if (response == NULL)
```
Comment (xunzhang) [inline src/backend/libpq/rangerrest.c]: Looks good 👋

File: src/backend/libpq/rangerrest.c [hunk 5]
```diff
@@ -24,60 +24,72 @@
...
+
+	struct json_object *response = json_tokener_parse(buffer);
+	if (response == NULL) 
+	{
+		elog(WARNING, "json_tokener_parse failed");
+		return RANGERCHECK_UNKNOWN;
+	}
+
+	struct json_object *accessObj = NULL;
+	if (!json_object_object_get_ex(response, "access", &accessObj))
```
Comment (xunzhang) [inline src/backend/libpq/rangerrest.c]: What about other fields check?

File: src/backend/libpq/rangerrest.c [hunk 6]
```diff
@@ -86,132 +98,132 @@ RangerACLResult parse_ranger_response(char* buffer)
...
-
-  json_object *jreqid = json_object_new_string("1");
-  json_object_object_add(jrequest, "requestId", jreqid);
-  json_object *jclientip = json_object_new_string("123.0.0.21");
-  json_object_object_add(jrequest, "clientIp", jclientip);
-  json_object *jcontext = json_object_new_string("SELECT * FROM DDDDDDD");
-  json_object_object_add(jrequest, "context", jcontext);
-
-  return jrequest;
+	json_object *juser = NULL;
```
Comment (xunzhang) [inline src/backend/libpq/rangerrest.c]: Here I think you'd better set the tab/space again to make the diff minimal for review. If it is tricky to do that, let it be.

**Comment by xunzhang** [comment]
@interma Could you please attach the compile warning message you fixed here?

**Comment by interma** [comment]
@xunzhang warning:

be-secure.c:322:1: warning: unused function 'report_commerror' [-Wunused-function]
report_commerror(const char *err_msg)
^
1 warning generated.

rangerrest.c:63:37: warning: 'json_object_object_get' is deprecated [-Wdeprecated-declarations]
    struct json_object *accessObj = json_object_object_get(response, "access");
                                    ^
/usr/local/include/json-c/json_object.h:290:56: note: 'json_object_object_get' has been explicitly marked deprecated here
THIS_FUNCTION_IS_DEPRECATED(extern struct json_object* json_object_object_get(struct json_object* obj,
                                                       ^
rangerrest.c:74:16: warning: 'json_object_object_get' is deprecated [-Wdeprecated-declarations]
      jallow = json_object_object_get(jvalue, "allowed");
               ^
/usr/local/include/json-c/json_object.h:290:56: note: 'json_object_object_get' has been explicitly marked deprecated here
THIS_FUNCTION_IS_DEPRECATED(extern struct json_object* json_object_object_get(struct json_object* obj,
                                                       ^
rangerrest.c:406:30: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
    appendStringInfo(&tname, rps_addr_host);
                             ^~~~~~~~~~~~~
rangerrest.c:406:30: note: treat the string as an argument to avoid this
    appendStringInfo(&tname, rps_addr_host);
                             ^
                             "%s",
3 warnings generated.

File: src/backend/libpq/rangerrest.c [hunk 7]
```diff
@@ -245,279 +257,326 @@ json_object *create_ranger_request_json_batch(List *args)
...
+
+	/* init the curl session */
+	curl_handle->curl_handle = curl_easy_init();
+	if (curl_handle->curl_handle == NULL)
+	{
+		goto _exit;
+	}
+
+	// hard-coded timeout
+	curl_easy_setopt(curl_handle->curl_handle, CURLOPT_TIMEOUT, 30L);
```
Comment (stanlyxiang) [inline src/backend/libpq/rangerrest.c]: Currently we could define a macro to replace this magic number.  
We could investigate whether this timeout should be a guc laterly.

Comment (interma) [inline src/backend/libpq/rangerrest.c]: Agreed, hard-code timeout now and we wil investigate in future.

File: src/backend/libpq/rangerrest.c [hunk 8]
```diff
@@ -245,279 +257,326 @@ json_object *create_ranger_request_json_batch(List *args)
...
-    memset(&curl_context, 0, sizeof(struct curl_context_t));
+	struct curl_context_t curl_context;
+	memset(&curl_context, 0, sizeof(struct curl_context_t));
 
-    /* call GET method to send request*/
-    call_ranger_rest(&curl_context, request);
+	/* call GET method to send request*/
+	if (call_ranger_rest(&curl_context, request) < 0)
+	{
+		return RANGERCHECK_UNKNOWN;
```
Comment (zhangh43) [inline src/backend/libpq/rangerrest.c]: Here return RANGERCHECK_UNKNOWN has some problem.
check_privilege_from_ranger is call by pg_rangercheck which return AclResult. But RANGERCHECK_UNKNOWN is belong to RangerACLResult. We need to unify them.

Comment (zhangh43) [inline src/backend/libpq/rangerrest.c]: Here return RANGERCHECK_UNKNOWN has some problem.
check_privilege_from_ranger is call by pg_rangercheck which return AclResult. But RANGERCHECK_UNKNOWN is belong to RangerACLResult. We need to unify them.

**Comment by zhangh43** [comment]
Return RANGERCHECK_UNKNOWN has some problem.
check_privilege_from_ranger is call by pg_rangercheck which returns AclResult. But RANGERCHECK_UNKNOWN is belong to RangerACLResult. We need to unify them.

File: src/test/regress/checkinc.py (hunk 3 above)
Comment (interma) [inline src/test/regress/checkinc.py]: Yes, only in "make install". 
"make" is ok.

File: src/backend/libpq/rangerrest.c (hunk 8 above)
Comment (interma) [inline src/backend/libpq/rangerrest.c]: tks! I will modify it.

**Comment by interma** [comment]
Fixed review problems already.
Tks, guys!

**Comment by stanlyxiang** [comment]
+1

File: src/backend/libpq/rangerrest.c [hunk 9]
```diff
@@ -272,280 +286,300 @@ json_object *create_ranger_request_json_batch(List *args)
...
+	curl_global_init(CURL_GLOBAL_ALL);
+
+	/* init the curl session */
+	curl_handle->curl_handle = curl_easy_init();
+	if (curl_handle->curl_handle == NULL)
+	{
+		goto _exit;
+	}
+
+	// hard core timeout
```
Comment (linwen) [inline src/backend/libpq/rangerrest.c]: ／* timeout is 30 seconds *／

File: src/backend/libpq/rangerrest.c [hunk 10]
```diff
@@ -452,8 +448,8 @@ static int call_ranger_rest(CURL_HANDLE curl_handle, const char* request)
 		goto _exit;
 	}
 
-	// hard core timeout
-	curl_easy_setopt(curl_handle->curl_handle, CURLOPT_TIMEOUT, 15L);
+	// hard-coded timeout
```
Comment (linwen) [inline src/backend/libpq/rangerrest.c]: please use /**/ as comment.
/* hard-coded timeout is 30 seconds */

File: src/test/regress/checkinc.py (hunk 3 above)
Comment (xunzhang) [inline src/test/regress/checkinc.py]: Make sense since Travis CI is green with this commit. Good fix.

**Comment by interma** [comment]
I will open a new PR, closed this.
